safehttp/plugins/xsrf/xsrf.go

```diff
@@ -63,24 +56,52 @@ func Token(r *safehttp.IncomingRequest) (string, error) {
 	return tok.(string), nil
 }
 
+func addTokenCookie(w *safehttp.ResponseWriter) (*safehttp.Cookie, error) {
+	buf := make([]byte, 20)
+	_, err := rand.Read(buf)
+	if err != nil {
+		panic(fmt.Errorf("failed to generate entropy using crypto/rand/RandReader: %v", err))
+	}
+
+	c := safehttp.NewCookie(tokenCookie, base64.StdEncoding.EncodeToString(buf))
+	c.SetSameSite(safehttp.SameSiteStrictMode)
+	if err := w.SetCookie(c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
 // Before should be executed before directing the safehttp.IncomingRequest to
 // the handler to ensure it is not part of the Cross-Site Request
 // Forgery attack.
 //
+// On first user visit through a state preserving request (GET, HEAD or
+// OPTIONS), it sets a nonce-based cookie used to identify a user. This will be
+// used in the token generation and verification algorithm and is expected in
+// all subsequent incoming requests.
+//
 // In case of state changing requests (all except GET, HEAD and OPTIONS), it
 // checks for the presence of an XSRF token in the request and validates it
-// based on the user ID associated with the request.
+// based on the custom cookie.
 //
 // For authorized requests, it adds a cryptographically safe XSRF token to the
-// incoming request. It can be later extracted using Token.
+// incoming request. This can be later extracted using Token.
 func (i *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, cfg interface{}) safehttp.Result {
-	userID, err := i.Identifier.UserID(r)
+	needsValidation := !statePreservingMethods[r.Method()]
+	c, err := r.Cookie(tokenCookie)
 	if err != nil {
-		return w.ClientError(safehttp.StatusUnauthorized)
+		if needsValidation {
+			return w.ClientError(safehttp.StatusForbidden)
+		}
+		c, err = addTokenCookie(w)
+		if err != nil {
+			// An error is returned when the plugin fails to Set the Set-Cookie
+			// header in the response writer as this is a server misconfiguration.
+			return w.ServerError(safehttp.StatusInternalServerError)
+		}
 	}
 
-	actionID := r.Method() + " " + r.URL.Path()
-	needsValidation := !statePreservingMethods[r.Method()]
+	actionID := r.URL.Path()
```

Comment:
Why this change?

Reply:
Having the method there is error-prone since we could visit the same path using different HTTP methods. I discussed offline with @empijei who said that it's fine if the `actionID` is just the path.